src/models/node.ts

```diff
@@ -117,39 +141,73 @@ export const isNode = <N extends INodeBase, E extends IEdgeBase>(obj: any): obj
 };
 
 export class Node<N extends INodeBase, E extends IEdgeBase> implements INode<N, E> {
-  public readonly id: number;
-  public data: N;
-  public position: INodePosition;
-  public style: INodeStyle = {};
-  public state = GraphObjectState.NONE;
+  protected readonly _id: number;
+  protected _data: N;
+  protected _position: INodePosition;
+  protected _style: INodeStyle = {};
+  protected _state = GraphObjectState.NONE;
 
+  private readonly _listeners: IObserver[] = [];
   private readonly _inEdgesById: { [id: number]: IEdge<N, E> } = {};
   private readonly _outEdgesById: { [id: number]: IEdge<N, E> } = {};
   private readonly _onLoadedImage?: () => void;
 
-  constructor(data: INodeData<N>, settings?: Partial<INodeSettings>) {
-    this.id = data.data.id;
-    this.data = data.data;
-    this.position = { id: this.id };
+  constructor(data: INodeData<N>, settings?: Partial<INodeSettings>, listener?: IObserver) {
```

Comment:
Is it fine to use a single listener in the constructor here? From an OOP perspective, using a listener list seems more natural, but I feel that a single listener is clearer and easier to use.

Reply:
Actually, it would be best to add listeners as an array to the `INodeSettings` because those are the initial settings for the node. Having more than > 2 params in JS is making things harder to use (e.g. you want to send a listener with default settings that would look like: `(data, undefined, listener)` 🤮 :D 

You can see that `INodeSettings` already has some utility stuff like callback for onLoadedImage - which in the future should be moved to Listener - but then we need to add what kind of events you want to listen to.

Same applies to edges.

Reply:
Also having this partial settings make things much more cleaner from "reading the code" perspective:

E.g.
`new Node(data, undefined, this);` - hard to know what `this` is for?

vs

`new Node(data, { listeners: [this] })` - object gives you a key `listeners`.